doc/design/persistent_volumes_etcd_data.md

```diff
@@ -0,0 +1,136 @@
+
+##### Effect
+
+The pod will go in an unknown/failed state, the etcd cluster remains quorate. Etcd-operator will remove the member from the etcd cluster and add a new member to the cluster and creates a new pod with a new name.
+
+#### Case C
+
+##### Event
+
+One of these events happens:
+* a majority of etcd processes exits (processes crashes, killed etc...) 
+* k8s nodes scheduling a majority of etcd members pods reboots and come back before node controller pod eviction triggers
+
+##### Effect
+
+The pods will go in failed state, the etcd cluster becomes unquorate. Etcd-operator can only restore the cluster from a backup.
+
+#### Case D
+
+##### Event
+
+* k8s nodes executing a majority of etcd members pods becomes partitioned (network partitioning, goes down, etc...) and cannot talk to the api server, after some time the node controller will evict their pods.
+
+##### Effect
+
+The pods will go in an unknown/failed state and eventually be removed from the API server, the etcd cluster becomes unquorate. Etcd-operator can only restore the cluster from a backup.
+
+
+#### Reasons
+
+Currently etcd operator handles cases A and B but when a cluster becomes unquorate (cases C and D) the current option is to restore the cluster from a backup.
+
+User may desire consistency over availability and prefer to wait for the cluster the return in a quorate state instead of restoring from a backup (that will contain old data).
+
+This proposal aims to fix cases C and D permitting to the cluster the return in a quorate state without the need to restore from a backup.
+
+## Proposed Changes
+
+#### Changes
+
+The change to be effective and cover the above failure scenario is split in two parts.
+
+### Part 1
+
+- Add to the cluster spec PodPolicy an option to enable putting etcd data inside a Persistent Volume
+- If podPolicy has persistent volumes enabled, when creating a new member also create a Persistent Volume Claim and use it as the volume source for etcd-data
+- Handle etcd data persistent volume claims cleanup:
+  - set on the PVC an owner like already done with other resources.
+  - in the reconcile loop, remove PVCs that don't have a matching etcd member in the etcd cluster and their related pod doesn't exists in the k8s API
+  - garbage collect also PVCs
+
+#### Pod name to Persistent Volume Claim name mapping
+
+The pod and related PVC are mapped 1 to 1 so their names be easily generated and deducted. I.E. if the pod is called `etcd-0000` its related PVC name will be `etcd-0000-pvc`
+
+#### Results
+
+These changes will put the basis for initial persistent volume management
+
+#### Problems
+
+This first part won't fix case C and D.
+
+#### Proposed enhancements to Part 1 (@xiang90)
+
+- When using a PVC as volume source also set pod restart policy to `Always`
+
+This will fix case C and change behavior of case A and B. The pod won't go in a failed state but the kubelet will try to restart it indefinetly.
+
```

Comment:
@hongchaodeng @xiang90 Perhaps it's also important to note that setting restart policy to `Always` will work also without persistent volumes since the emptydir is removed only when the pod is deleted (https://kubernetes.io/docs/concepts/storage).

So changing restart policy to `Always` is making this proposal more long and confused (multiple combinations) and will probably create some undesired effects and increasing.

So I was thinking to remove it entirely from the design. What do you think?

Reply:
What if the member is removed? In that case we don't want restart.

Reply:
Just to give more context, when a pod is removed, it will exit 0. But it's also fine if it restart: it will keep failing until kubelet kills it.

I think moving forward we will have all pods to always restart. Let's keep it a separate topic and open a new issue for it.

Reply:
> I think moving forward we will have all pods to always restart. Let's keep it a separate topic and open a new issue for it.

I'm not sure if setting Restart Policy to Always is a good idea but you'll probably have other ideas to handle some cases masked by using it, so if you're going to open a new issue I'll add my opinions there.

> Let's keep it a separate topic 

Good. Will remove all the parts related to changing the current pod Restart Policy.